src/apps/replication/lib/replica_2pc.cpp

```diff
@@ -290,12 +292,15 @@ void replica::on_prepare(dsn_message_t request)
     auto mu2 = _prepare_list->get_mutation_by_decree(decree);
     if (mu2 != nullptr && mu2->data.header.ballot == mu->data.header.ballot)
     {
-        ddebug( "%s: mutation %s redundant prepare skipped", name(), mu->name());
-
         if (mu2->is_logged())
         {
             ack_prepare_message(ERR_OK, mu);
         }
+        else
+        {
+            derror("%s: mutation %s on_prepare skipped as it is duplicate", name(), mu->name());
+            // TODO(qinzuoyan): response something or do merge response
```

Comment:
this will be unnecessary when we add retry logic in rpc engine - the retried rpc will use the same id therefore it will be considered responded even the response is for a previous try.